feat!: Docker CI - Security added & Tagging updates - 2

jrmanes opened the pull request:
## Overview

ℹ️ Same content as: [1320](https://github.com/celestiaorg/celestia-app/pull/1320), I had an issue and the PR was closed...

Hello team, 

Hope you’re doing well 

The following PR contains some changes in the Docker CI, build & publish.

I’ve added some features:
- File renamed: `amd64-docker-build.yml` -> `docker-build-publish.yml`
*We will be able to use a matrix for building the container in different architectures.*

- This CI is triggered when
  - When push to any branch
  - When creating `PRs`
  - When there's a new hotfix (when push to main branch)
  - When push tags `(v0.0.0/v0.0.0-alpha/v0.0.0-beta/v0.0.0-rcX)`
  - Only push new Docker images when they are in `main` or `v*` (blocking forks for security)

- Security: 
[Trivy](https://www.aquasec.com/products/trivy/) added to the CI, this is the first step during the build, it scans the image and provide us a table summary in case we have any CVE, if everything goes well, it continues with the next step.
*It’s not going to block the CI in case of bugs, though I think it will be good for us, stop it and fix them.*

- Docker tagging:
  - Git short SHA => for example: `a594b69`
  - Tag ID => for example: `0.0.1`
- Provided some additional metadata to the images:
  - Maintainer  =>   "maintainer": "CelestiaOrg"
  - Description => "CelestiaOrg repository celestiaorg/celestia-app"
  - URL to the specific commit =>  "commit_url": "https://github.com/celestiaorg/celestia-app/commit/a594b691ac019fc39afc7ee28c2798269ef0b159"
  - Docker pull command => "docker_pull_command": "docker pull ghcr.io/celestiaorg/celestia-app:a594b691"

---

## Checklist

- [x] Required CI checks are passing
- [x] Linked issues closed with keywords


---

## Blockers

Hello team!
I'll need to add some permissions to allow the CI to have the access to the packages.

This is an error that I'm having: [link](https://github.com/celestiaorg/celestia-app/actions/runs/4058197292/jobs/6984898432)

```
ERROR: failed to solve: failed to push ghcr.io/celestiaorg/celestia-app:pr-1320: unexpected status: 403 Forbidden
Error: buildx failed with: ERROR: failed to solve: failed to push ghcr.io/celestiaorg/celestia-app:pr-1320: unexpected status: 403 Forbidden
```

Thanks in advance!

cc: @evan-forbes @rootulp  @Bidon15 @sysrex 

Please, ping me when you'll going to merge it, just to check that everything goes fine 😊 
Thank you team!

--- 
Closes Issue: [37](https://github.com/celestiaorg/devops/issues/37)

SweeXordious commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,100 @@
+      - "v[0-9]+.[0-9]+.[0-9]+-alpha.[0-9]+"
+      - "v[0-9]+.[0-9]+.[0-9]+-beta.[0-9]+"
+      - "v[0-9]+.[0-9]+.[0-9]+-rc[0-9]+"
+  pull_request:
```
[Question]
Won't building and publishing docker images for every branch be too much?

jrmanes replied:
I think so, though after discuss it with the team, it's okay to keep a track for all the images (probably, we should clean them after X time in the future) as we currently are doing: [packages](https://github.com/celestiaorg/celestia-app/pkgs/container/celestia-app/versions?filters%5Bversion_type%5D=tagged).

SweeXordious replied:
I see, thanks

SweeXordious reviewed (commented):
:+1: :+1:

Can you resolve the conflicts please?

Bidon15 commented on .github/workflows/amd64-docker-build.yml:
```diff
@@ -37,7 +93,8 @@ jobs:
           push: true
           tags: ${{ steps.meta.outputs.tags }}
-          file: Dockerfile
+          labels: ${{ steps.meta.outputs.labels }}
```
```suggestion
          labels: ${{ steps.meta.outputs.labels }}

```

jrmanes replied:
thank you! @Bidon15 😊

evan-forbes reviewed (changes requested):
> buildx failed with: ERROR: failed to solve: failed to push ghcr.io/celestiaorg/celestia-app:pr-1327: unexpected status: 403 Forbidden

are we still seeing the issues that we saw with #1320 with the failed docker build CI? 

besides fixing the linters, this LGTM

MSevey commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,131 @@
+          tags: ${{ steps.meta.outputs.tags }}
+          file: Dockerfile
+
+  # docker-security:
```
why are we keeping all this commented out code?

jrmanes replied:
@evan-forbes and me discuss about the `403` error, and we thought about to comment step by step and see which part was causing the issues, atm is using the current version (the old one), but once we fix the issue, this will be uncommented

evan-forbes replied:
we're just debugging the CI atm to try and figure out which thing is causing the docker permissions to flip out.

evan-forbes commented:
@jrmanes suggested me trying this since PR since I write permissions, so I'm closing here for now to let me open up a PR from the same branch

jrmanes commented:
hello @MSevey  @Bidon15 

I've added the changes that we discuss last Friday for fixing the issues with the forks and so far the CI looks better.
- Split the build & publish 
  - **build**: always, it's ok
  - **publish**: only when the branch is `main` or when releases `v*`
  
 Let me know if you find anything that you would like to add :) 
 
 thanks in advance!

Bidon15 commented:
> publish: only when the branch is main or when releases v*

Does it mean that latest tag is always == to what is merged on main, right? @jrmanes 
cc: @smuu

jrmanes commented:
> 

Yes, we'll build it alway, but **only** push the new images when the branch is `main` or `v*`, meaning that the tag `latest` will be updated during that process.
We have to test it when we'll merge it, but it should works in that way

jrmanes commented:
> > 
> 
> Yes, we'll build it alway, but **only** push the new images when the branch is `main` or `v*`, meaning that the tag `latest` will be updated during that process. We have to test it when we'll merge it, but it should works in that way

Update: 
After discuss it:
- We postpone the `latest` topic for future

MSevey commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,123 @@
+        with:
+          push: false
+          platforms: linux/amd64
+          tags: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:latest
```
If I understand this correctly, we need to identify a tag in order to reference it in the next step correct?

Since we are not pushing anything, this tag is only every used as a local reference?

If that is the case, can you add a comment to document that? Maybe we could even change the tag to `local` or something to further avoid potential confusion about defining a `latest` image.

jrmanes replied:
Exactly @MSevey 
We're building the container prior the scan.
I like the idea to change the tag to `local` instead of `latest`👌

MSevey commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,123 @@
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build Docker Image
+        uses: docker/build-push-action@v3
+        with:
+          platforms: linux/amd64
+          push: false
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          file: Dockerfile
+
+      - name: Push Docker Image
+        uses: docker/build-push-action@v3
+        if: |
+          github.ref == 'refs/heads/main' ||
+          startsWith(github.ref, 'refs/tags/v')
+        with:
+          platforms: linux/amd64
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          file: Dockerfile
```
Since we are using the combined build/push action, we can just set the push flag dynamically to avoid building twice.

```suggestion
      # We always build the image but we only push if we are on the `main` branch or a versioned `v*` branch
      - name: Build and PushDocker Image
        uses: docker/build-push-action@v3
        with:
          platforms: linux/amd64
          push: github.ref == 'refs/heads/main' || startsWith(github.ref, 'refs/tags/v')
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}
          file: Dockerfile
```

jrmanes replied:
Love this idea! I hadn't seen this kind of `if` inline, looks interesting 👀

MSevey commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,123 @@
+            # output minimal (short sha)
+            type=raw,value={{sha}}
+            # output 0.2.1
+            type=semver,pattern={{version}}
```
where are we pulling in the version from?

jrmanes replied:
If I'm not mistaken, this is inherit from the Github Action: `docker/metadata-action@v4` (in the same step: `Extract Docker Metadata`)

https://github.com/docker/metadata-action#typesemver
In order to generate the format: `1.2.3`

Bidon15 replied:
ser, forgot to chime in 

can we use `v{{version}}`?  

I don't want devs to be annoyed by remembering to remove the `v` from pulling the image release tag. Motivation is behind what we get used to in go.mod with `v`s

jrmanes replied:
of course! I personally prefer to have it 👌

MSevey reviewed (changes requested):
A few last questions

smuu commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,116 @@
+        uses: docker/build-push-action@v3
+        with:
+          push: false
+          platforms: linux/amd64
```
Do we also want to build it for arm64? That would support starting containers on mac and RPIs.

jrmanes replied:
I think it worth it and cost nothing to add ;)

Bidon15 reviewed (dismissed):
LGTM

jrmanes commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,116 @@
+        uses: docker/build-push-action@v3
+        with:
+          platforms: linux/amd64,linux/arm64
+          push: ${{ github.ref == 'refs/heads/main' || startsWith(github.ref, 'refs/tags/v') }}
```
please 🙏 hahha that would be great!

smuu commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,118 @@
+          # refer to it later
+          tags: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:local
+
+      - name: Run Trivy vulnerability scanner
```
Do we need to ensure that trivy is scanning both amd and arm? Or can we safely assume that the alpine image has the same vulnerabilities in amd and arm?

jrmanes replied:
for now, I think we can keep with `amd64` at least as an initial iteration, but will be a good point to have in future, also, I'm not sure if the Github action itself can do it, I cannot find the platform input... it's possible that we could tag both images and scanning them independently

MSevey commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,118 @@
+      - "v[0-9]+.[0-9]+.[0-9]+-beta.[0-9]+"
+      - "v[0-9]+.[0-9]+.[0-9]+-rc[0-9]+"
+  pull_request:
+  workflow_dispatch:
+    # Inputs the workflow accepts.
+    inputs:
+      version:
+        # Friendly description to be shown in the UI instead of 'name'
+        description: "Semver type of new version (major / minor / patch)"
+        # Input has to be provided for the workflow to run
+        required: true
+        type: choice
+        options:
+          - patch
+          - minor
+          - major
```
```suggestion
```
[non-blocking] This can be removed as it isn't being used.

jrmanes replied:
cleaned up 🧹

MSevey reviewed (dismissed):
LGTM now. We already discussed the follow ups of refactoring this out into the organizational level `.github` to apply to other repos. At that point we can move the trigger into ci_release workflow so that we have the full flow in place. Additionally that will allow us to block for linting and tests, since if either of those fail, there is no need to build a docker image.

smuu reviewed (approved):
LGTM

smuu commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,107 @@
+          tags: |
+            # output minimal (short sha)
+            type=raw,value={{sha}}
+            # output 0.2.1
```
[non-blocking] Comment does not include 'v' as prefix

# output 0.2.1 -> # output v0.2.1

evan-forbes commented on .github/workflows/docker-build-publish.yml:
```diff
@@ -0,0 +1,107 @@
+          # refer to it later
+          tags: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:local
+
+      - name: Run Trivy vulnerability scanner
```
can we run this locally if it fails?

smuu replied:
https://github.com/aquasecurity/trivy#installation
